libc/spec/gpu_ext.td

```diff
@@ -1,15 +1,41 @@
+def RPCPortT : NamedType<"rpc_port_t">;
+def RPCPortPtrT : PtrType<RPCPortT>;
+def RPCBufferT : NamedType<"rpc_buffer_t">;
+def RPCBufferPtrT : PtrType<RPCBufferT>;
+def RPCCallbackT : NamedType<"rpc_callback_t">;
+
 def GPUExtensions : StandardSpec<"GPUExtensions"> {
   HeaderSpec RPC = HeaderSpec<
     "gpu/rpc.h",
     [], // Macros
-    [], // Types
+    [RPCPortT, RPCBufferT,  RPCCallbackT, SizeTType], // Types
     [], // Enumerations
     [
         FunctionSpec<
             "rpc_host_call",
             RetValSpec<VoidType>,
             [ArgSpec<VoidPtr>, ArgSpec<VoidPtr>, ArgSpec<SizeTType>]
         >,
+        FunctionSpec<
+            "rpc_open_port",
+            RetValSpec<RPCPortT>,
+            [ArgSpec<UnsignedIntType>]
+        >,
+        FunctionSpec<
+            "rpc_send",
+            RetValSpec<VoidType>,
```

Comment:
design question: Would it make sense to give send, recv, and close return values of some sort? Probably an int so it can return various error codes or 0 on success to match other libc functions.

Reply:
I don't think that's necessary, there shouldn't be any failure modes once a port has been opened. Because we are executing in a unified address space, we don't need to worry about network effects like gRPC or similar. Internally we pretty much just spin until memory happens and then continue. We could potentially expand the internface to allow the lambda we're copying to fail, but I don't think there's any utility in that. If it's for consistently I could make it return an int and just hardwire it to zero.

Reply:
No error codes. It is fundamentally crucial that these functions do not and can not fail.

This permits implementing arbitrary functions that execute across the two machines with exactly the interface they have on one machine. No extra integer value to mean "network failed", no replacing double return types with a pair of double and boolean, no extra global variable named rpcerrno that people are supposed to check.

This trickery relative to the usual RPC problem of messing up the interfaces works because communication on a shared memory machine only fails when the whole machine has fallen over, which is non-recoverable, and each function does not dynamically allocate a resource that can fail.

At least by design - in practice we may currently be undersizing the ports structure to guarantee progress in the worst case scheduling, but if so that's fixable by making N large enough for a given machine. The cost is order of a few megabytes of memory which in the scheme of things doesn't matter for a singleton.